pyccel/ast/numpyext.py

```diff
@@ -786,25 +786,44 @@ def __new__(cls, a, dtype=None, order='K', subok=True, shape=None):
 class NumpyNorm(PyccelInternalFunction):
     """ Represents call to numpy.norm"""
     _dtype = NativeReal()
-
-    def __init__(self, arg, dim=None):
-        if isinstance(dim, ValuedArgument):
-            dim = dim.value
-        if self.dim is not None:
-            sh = list(sh)
-            del sh[self.dim]
+    _shape = ()
+    _rank  = 0
+
+    def __init__(self, arg, axis=None):
+        super().__init__(arg, axis)
+        cmplx_precision = None
+        if isinstance(arg.dtype, NativeBool):
+            arg = PythonFloat(arg)
+        elif isinstance(arg.dtype, NativeComplex):
+            cmplx_precision = arg.precision
+            arg = PythonAbs(arg)
```

Comment:
As already discussed I think the call to  `PythonAbs` should be added in fcode.py as other languages may implement a norm for complexes. This should be used if possible as it removes unnecessary sqrt calls.

E.g. In a language with cnorm, the complex norm is calculated as:
```
sqrt( a.real**2 + a.imag**2 + b.real**2 + b.imag**2 + ...)
```
but in a language without a complex norm it must be calculated as:
```
sqrt( sqrt(a.real**2 + a.imag**2)**2 + sqrt(b.real**2 + b.imag**2)**2 + ...)
```

Reply:
i though we should keep it in the ast since numpy is doing it internally.

Reply:
The question is whether numpy is the outlier here, or fortran. C obviously does neither as it doesn't handle arrays. I wonder what julia does.

That said I suspect that when we implement it in c we will create a `norm` and a `cnorm` function call as this will avoid the unnecessary `sqrt**2` and will therefore run faster. If that is the case then the `abs` will cause problems